### PR TITLE
Add `generator=` kwarg for DataLoader & random samplers

### LINKED ISSUE #39572: Required some argument in dataloader for setting randomstate
## Motivation
I am really confused regarding how to set random state in dataloader in PyTorch. I had searched PyTorch forums as well as all related issues but nothing works for me. If there is a simple argument regarding the random state of dataloader then it will be very helpful.

## Pitch

I am working on a project in which I need to load image data from directories. All worked fine but as soon as I executed tried to execute the dataloader again then the results were different. I set shuffle to True but I want some random state in dataloader argument similar to that of train_test_split function in scikit-learn library
## Additional context

cc @SsnL

### PR DESCRIPTION
**Pull request by ShawnZhong**
Fix #39572

Add `generator=` kwarg for DataLoader & random samplers

cc: @SsnL, @deeppatel4557, @albanD, @mitar

### CODE REVIEW

**Review by ssnl — CHANGES_REQUESTED** [review]
the generator for DataLoader should also be used for `base_seed` in iterators

**Review by ssnl — COMMENTED** [review]
can we test the base_seed reproducibility by doing something similar to `test_worker_seed`?

File: test/test_dataloader.py [hunk 1]
```diff
@@ -1203,6 +1203,15 @@ def test_worker_seed(self):
             seeds.add(batch[0])
         self.assertEqual(len(seeds), num_workers)
 
+    def test_worker_seed_reproducibility(self):
+        num_workers = 6
+        batch_size = 1
+        dataset = SynchronizedSeedDataset(num_workers, batch_size, num_workers)
+        dataloader = DataLoader(dataset, batch_size=batch_size, num_workers=num_workers, generator=torch.Generator().manual_seed(42))
+        actual_seeds = set(int(batch) for batch in dataloader)
+        expected_seeds = set(6909045637428952499 + i for i in range(num_workers))
```
Comment (SsnL) [inline test/test_dataloader.py:1212]: don't do this. seeding is only guaranteed to be consistent on the same device & version. just do two runs and compare results

Comment (ShawnZhong) [reply test/test_dataloader.py:1212]: 👌

**Review by ssnl — APPROVED** [review]
merge when green. thanks!

**Comment by albanD** [comment]
Windows CI is having a big hiccup because of some `pip` issues. Don't worry about those :)

**Review by albanD — APPROVED** [review]
Thanks for the PR.

**Comment by SsnL** [comment]
Hi @ShawnZhong , do you mind also updating the notes at https://pytorch.org/docs/stable/data.html#randomness-in-multi-process-data-loading https://pytorch.org/docs/stable/notes/faq.html#my-data-loader-workers-return-identical-random-numbers about this too? Thanks :)
